examples/relu-backprop-compile/tests/test_relu.py

```diff
@@ -0,0 +1,158 @@
+    assert x.grad is not None
+    assert x.grad.shape == x.shape
+
+    expected_grad = torch.where(
+        x > 0,
+        torch.tensor(1.0, dtype=dtype, device=device),
+        torch.tensor(0.0, dtype=dtype, device=device),
+    )
+    torch.testing.assert_close(x.grad, expected_grad)
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
+def test_relu_gradient_comparison(dtype):
+    device = get_device()
+    x_data = torch.tensor(
+        [[-2.0, -1.0, 0.0, 1.0, 2.0], [0.5, -0.5, 1.5, -1.5, 0.0]],
+        dtype=dtype,
+        device=device,
+    )
+
+    x_kernel = x_data.clone().requires_grad_(True)
+    y_kernel = relu.relu(x_kernel)
+    loss_custom = y_kernel.sum()
+    loss_custom.backward()
+
+    x_torch = x_data.clone().requires_grad_(True)
+    y_torch = F.relu(x_torch)
+    loss_torch = y_torch.sum()
+    loss_torch.backward()
+
+    torch.testing.assert_close(y_kernel, y_torch)
+    torch.testing.assert_close(x_kernel.grad, x_torch.grad)
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
+def test_relu_backward_chain(dtype):
+    device = get_device()
+    x = torch.randn(64, 128, dtype=dtype, device=device, requires_grad=True)
+    y = relu.relu(x)
+    z = y * 2.0
+    loss = z.sum()
+    loss.backward()
+
+    assert x.grad is not None
+    assert x.grad.shape == x.shape
+
+    expected_grad = torch.where(
+        x > 0,
+        torch.tensor(2.0, dtype=dtype, device=device),
+        torch.tensor(0.0, dtype=dtype, device=device),
+    )
+    torch.testing.assert_close(x.grad, expected_grad)
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize(
+    "shape",
+    [
+        (32, 64),
+        (1, 1024),
+        (128, 256),
+        (8, 8, 16),
+    ],
+)
+def test_relu_fwd_opcheck(shape, dtype):
+    device = get_device()
+    x = torch.randn(shape, dtype=dtype, device=device, requires_grad=True)
+    opcheck(relu.ops.relu_fwd, (x,))
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
+def test_relu_torch_compile(dtype):
+    device = get_device()
+
+    class SimpleModel(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.linear = torch.nn.Linear(1024, 1024)
+
+        def forward(self, x):
+            return relu.relu(self.linear(x))
+
+    model = SimpleModel().to(device).to(dtype)
+    compiled_model = torch.compile(model, fullgraph=True)
+
+    x = torch.randn((1024, 1024), dtype=dtype, device=device, requires_grad=True)
+
+    y_original = model(x)
+    y_compiled = compiled_model(x)
+
+    torch.testing.assert_close(y_original, y_compiled)
+
+    loss_original = y_original.sum()
+    loss_compiled = y_compiled.sum()
+
+    if x.grad is not None:
+        x.grad.zero_()
+
+    loss_original.backward(retain_graph=True)
+    assert x.grad is not None
+    grad_original = x.grad.clone()
+
+    x.grad.zero_()
+    loss_compiled.backward()
+    assert x.grad is not None
+    grad_compiled = x.grad.clone()
+
+    torch.testing.assert_close(grad_original, grad_compiled)
```

Comment:
Another nice thing to check is that this compiled model (with fullgraph traceability) doesn't trigger recompilations. It can be checked like so:
https://github.com/huggingface/diffusers/blob/764b62473ac10afd9c52e6b3f3f528f719bc7a34/tests/models/test_modeling_common.py#L2067